x/evm/keeper/grpc_query.go

```diff
@@ -336,6 +333,34 @@ func (k Keeper) EstimateGasInternal(c context.Context, req *types.EthCallRequest
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
+	// Recap the highest gas limit with account's available balance.
+	if msg.GasFeeCap().BitLen() != 0 {
+		balance := k.bankKeeper.GetBalance(ctx, sdk.AccAddress(args.From.Bytes()), k.GetParams(ctx).EvmDenom)
+		available := balance.Amount
+		if args.Value != nil {
+			if args.Value.ToInt().Cmp(available.BigInt()) >= 0 {
+				return nil, core.ErrInsufficientFundsForTransfer
+			}
+			available = available.Sub(sdkmath.NewIntFromBigInt(args.Value.ToInt()))
+		}
+		allowance := available.Quo(sdkmath.NewIntFromBigInt(msg.GasFeeCap()))
+
+		// If the allowance is larger than maximum uint64, skip checking
+		if allowance.IsUint64() && hi > allowance.Uint64() {
+			transfer := "0"
+			if args.Value != nil {
```

Comment:
Nit: we're checking twice the `args.Value != nil` and the `transfer` string is only used on the log.
Can we refactor this to set `transfer := "0"` before the first nil check and update it inside it?
So we can remove the second nil check